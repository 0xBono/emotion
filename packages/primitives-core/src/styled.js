// @flow
import * as React from 'react'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign,
  interleave
} from './utils'
import { withCSSContext } from '@emotion/core'
import { createCss } from './css'

let defaultPickTest = prop => prop !== 'theme' && prop !== 'innerRef'

type options = {
  getShouldForwardProp: (cmp: React.ElementType) => (prop: string) => boolean
}

export function createStyled(
  StyleSheet: Object,
  { getShouldForwardProp = () => defaultPickTest }: options = {}
) {
  const css = createCss(StyleSheet)

  return function createEmotion(component: React.ElementType) {
    let pickTest = getShouldForwardProp(component)

    return function createStyledComponent(...rawStyles: *) {
      let styles

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles)
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      let Styled = withCSSContext((props, context) => {
        let mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: props.theme || context.theme
        })
        let stylesWithStyleProp = styles
        if (props.style) {
          stylesWithStyleProp = styles.concat(props.style)
        }
        const emotionStyles = css.apply(mergedProps, stylesWithStyleProp)

        return React.createElement(
          component,
          pickAssign(pickTest, {}, props, {
            ref: props.innerRef,
            style: emotionStyles
          })
        )
      })
      // $FlowFixMe
      Styled.withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      Styled.displayName = `emotion(${getDisplayName(component)})`

      return Styled
    }
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
