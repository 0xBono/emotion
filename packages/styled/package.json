{
  "name": "@emotion/styled",
  "version": "0.10.6",
  "description": "styled API for emotion",
  "main": "dist/styled.cjs.js",
  "module": "dist/styled.esm.js",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/emotion/tree/master/next-packages/styled",
  "dependencies": {
    "@emotion/styled-base": "^0.10.6",
    "babel-plugin-emotion": "^9.2.10"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "src",
    "dist",
    "macro.js"
  ],
  "umd:main": "./dist/index.min.js",
  "browser": {
    "./dist/styled.cjs.js": "./dist/styled.browser.cjs.js",
    "./dist/styled.esm.js": "./dist/styled.browser.esm.js"
  }
}
