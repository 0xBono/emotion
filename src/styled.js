import React from 'react'
import { css } from './index'

export default function (tag, cls, vars = [], content) {
  function Styled (props) {
    const className = css(
      cls,
      vars.map(v => (v && typeof v === 'function' ? v(props) : v)),
      content
    )

    return React.createElement(
      tag,
      Object.assign({}, props, {
        className: props.className
          ? props.className +
          ' ' +
          className
          : className
      })
    )
  }

  Styled.displayName = `Styled[${typeof tag === 'string' ? tag : tag.displayName || 'Wrapped'}](${cls})`

  return Styled
}
