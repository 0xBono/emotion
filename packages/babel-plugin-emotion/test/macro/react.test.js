// @flow
import 'test-utils/legacy-env'
import React from 'react'
import renderer from 'react-test-renderer'
import styled from 'react-emotion/macro'

test('no dynamic', () => {
  const H1 = styled.h1`
    float: left;
  `

  const tree = renderer.create(<H1>hello world</H1>).toJSON()

  expect(tree).toMatchSnapshot()
})

test('basic render', () => {
  const fontSize = 20
  const H1 = styled.h1`
    color: blue;
    font-size: ${fontSize + 'px'};
    @media (min-width: 420px) {
      color: blue;
    }
  `

  const tree = renderer.create(<H1>hello world</H1>).toJSON()

  expect(tree).toMatchSnapshot()
})

test('basic render with object as style', () => {
  const fontSize = 20
  const H1 = styled.h1({ fontSize })

  const tree = renderer.create(<H1>hello world</H1>).toJSON()

  expect(tree).toMatchSnapshot()
})

test('object as style', () => {
  const H1 = styled.h1(
    props => ({
      fontSize: props.fontSize
    }),
    props => ({ flex: props.flex }),
    { display: 'flex' }
  )

  const tree = renderer
    .create(
      <H1 fontSize={20} flex={1}>
        hello world
      </H1>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
