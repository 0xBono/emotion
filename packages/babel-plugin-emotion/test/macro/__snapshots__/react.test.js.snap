// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic render 1`] = `
.emotion-0 {
  color: blue;
  font-size: 20px;
}

@media (min-width:420px) {
  .emotion-0 {
    color: blue;
  }
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`basic render with object as style 1`] = `
.emotion-0 {
  font-size: 20px;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`no dynamic 1`] = `
.emotion-0 {
  float: left;
}

<h1
  className="emotion-0 emotion-1"
>
  hello world
</h1>
`;

exports[`object as style 1`] = `
.emotion-0 {
  font-size: 20px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<h1
  className="emotion-0 emotion-1"
  fontSize={20}
>
  hello world
</h1>
`;
